src/audio/asrc/IasSrcFarrow_generic.c

```diff
@@ -0,0 +1,386 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2012 Intel Corporation. All rights reserved.
+
+#include <sof/audio/asrc/asrc_config.h>
+
+#if ASRC_GENERIC == 1
+
+#include <sof/audio/asrc/IasSrcFarrow.h>
+#include <sof/audio/format.h>
+
+void firFilter16(struct tagIasSrcFarrow *srcObj, int16_t **outputBuffers,
+		 uint32_t indexOutputFrame)
+{
+	int64_t prod;
+	int32_t prod32;
+	int16_t prod16;
+	int32_t *filterP;
+	int16_t *bufferP;
+	int indexChannel;
+	int indexFilter;
+	int indexFilterLimit;
+
+	/*
+	 *  Four filter bins are accumulated per iteration.
+	 * 'indexFilterLimit' is therefore stored to reduce redundant
+	 * calculations.
+	 */
+	indexFilterLimit = srcObj->mFilterLength >> 2;
+
+	/* Iterate over each channel */
+	for (indexChannel = 0; indexChannel < srcObj->mNumChannels;
+	     indexChannel++) {
+		/* Pointer to the beginning of the impulse response */
+		filterP = &srcObj->mImpulseResponse[0];
+
+		/* Pointer to the buffered input data */
+		bufferP = &srcObj->mRingBuffers16[indexChannel]
+			[srcObj->mBufferWritePosition];
+
+		/* Initialise the accumulator */
+		prod = 0;
+
+		/* Iterate over the filter bins.
+		 * Data is Q1.15, coefficients are Q1.30. Prod will be Qx.45.
+		 */
+		for (indexFilter = 0; indexFilter < indexFilterLimit;
+		     indexFilter++) {
+			/* Multiply and accumulate */
+			prod += (int64_t)(*bufferP--) * (*filterP++);
+			prod += (int64_t)(*bufferP--) * (*filterP++);
+			prod += (int64_t)(*bufferP--) * (*filterP++);
+			prod += (int64_t)(*bufferP--) * (*filterP++);
```

Comment:
I kept the same loop unroll by four in C version as in HiFi3 optimized but I have no evidence of improved performance. To save code this could be reverted to no unroll. I need to give anyway another shot for plain C version if I could achieve real-time operation. Currently it's too slow for real-time.